File: leetcode/0410-split-array-largest-sum/solution.py
from typing import List


class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        def feasible(cap: int) -> bool:
            count, cur = 1, 0
            for x in nums:
                if cur + x > cap:
                    count += 1
                    cur = x
                    if count > m:
                        return False
                else:
                    cur += x
            return True

        lo, hi = max(nums), sum(nums)
        while lo < hi:
            mid = (lo + hi) // 2
            if feasible(mid):
                hi = mid
            else:
                lo = mid + 1
        return lo

File: leetcode/0410-split-array-largest-sum/Solution.java
class Solution {
    public int splitArray(int[] nums, int m) {
        long lo = 0, hi = 0;
        for (int x : nums) {
            lo = Math.max(lo, x);
            hi += x;
        }

        while (lo < hi) {
            long mid = lo + (hi - lo) / 2;
            if (feasible(nums, m, mid)) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return (int) lo;
    }

    private boolean feasible(int[] nums, int m, long cap) {
        int count = 1;
        long cur = 0;
        for (int x : nums) {
            if (cur + x > cap) {
                count++;
                cur = x;
                if (count > m) return false;
            } else {
                cur += x;
            }
        }
        return true;
    }
}
